packages/store/addon/-private/ts-interfaces/identifier.ts

```diff
@@ -194,7 +200,15 @@ export type GenerationMethod = (data: Object, bucket: string) => string;
   @public
   @static
 */
-export type UpdateMethod = (identifier: StableIdentifier, newData: Object, bucket: string) => void;
+export type UpdateMethod = {
+  (identifier: StableRecordIdentifier, newData: ResourceData, bucket: 'record'): void;
+  (identifier: StableIdentifier, newData: unknown, bucket: never): void;
+  (
+    identifier: StableIdentifier | StableRecordIdentifier,
+    newData: ResourceData | unknown,
+    bucket: IdentifierBucket
+  ): void;
+};
```

Comment:
For the second overload, I *think* we want to leave off `bucket` rather than setting it as `never`. If you include the param but set it to never, callers invoking with that signature will get an error if they only pass two arguments. Additionally, I *think* you don't need the final overload at all here – only at the implementation site(s).

```suggestion
export type UpdateMethodAlt = {
  (identifier: StableIdentifier, newData: unknown): void;
  (identifier: StableRecordIdentifier, newData: ResourceData, bucket: 'record'): void;
};
```

See [playground](https://www.typescriptlang.org/play?#code/PTAEElQCwQwNwKagHYHtQEsAmCagO6wAuoRUSGyRCATsjADYDOoqAZqeU0jDQgDSgm6SACsArkxIBbGDk4JpoPgxjUAUCFAwADjpqp9GNQgYBPUACMY3LK2QFiEUACosqF6ADWafNsuo4grSAHSgAFyiqGbh6uqU1DRsMADGSADKRDCWDAjgOFQYbBi0oADeoCmoqAwAQrjITOGgAORmqC2gAL4A3PFUtMlpoJnZuQBKCFU0WPkIhcWlCAAe1MhYLKM5eQVERSU05Y4miDTNbbhQnb3qRGY6SJPC4jRpACJqeAC8RwD6vzkYFBmkQaOIkDc7g8ILt9rRauIUl4ECQflIaJQAOZ9dQrHSoGgkKFIACqOiwJgAsiioKg7D8yupQKAABTYeZ7RZnEZZbaTaazWFcwTIBD4D5ZZpPQKvBASmCCSyI5FEc58AUtACUzTgqGwfWZbKFB2aW1ycwWBxFYvlzXEyB8qHwyEVypRzVFp21oF1+qZrP9zPZltopt55uNpQAPjyxgh+QTBRy4TR+IGUDbPlKEM9ZfLQDH7Y7nWnmcylUj3TDk1yEZWiP7vb6sH0buocClVHxQLkSMtmmSKdRqWQ6X1liyKlUavUYI1mgAidoL7qCFqEE60LU4vEEon3Unkqk0ukAQQYqPK-qNNZNse2Fs5Voz4qzoCLvmQTb1Levwafob3hMUyJo+KbWq+kqgNKLzvJ8rr1mqIEzFqOo-q2OIdl2SC9qAMSgIOx6jlg55EH0ZiTpU1R1A0TSgEuqArl0a4btQpzbnEuEAF4Dkew4niRF6gD8N4htyZo7Le0ZAfGyFJmJEG2u+DqfghKoAPxIRq37YMJAB8V5lpgHAsuuxAIOxmAOP+KaaoZRllloITOemXTqG5QA) for a demo that this simplification works *if* I understand the intent correctly.

Reply:
Is already noticed we didn't need the implementation version within the interface and could leave only the overloads, so good call I'll remove it.

The thing about the second overload though is it is intended to be an error if invoked. It forces the user to implement both a signature and a JS implementation that can handle extension later. I'll demonstrate this to you if you'd like, maybe we can firm it up a little.

Reply:
We discussed this out of band and the two overloads are the desired overloads, as this is an interface for a user implemented function and the second overload is serving as a placeholder to ensure the implemented function is compatible with us extending this interface with additional overloads later.